src/librustc_mir/hair/pattern/_match.rs

```diff
@@ -581,6 +616,143 @@ fn max_slice_length<'p, 'a: 'p, 'tcx: 'a, I>(
     cmp::max(max_fixed_len + 1, max_prefix_len + max_suffix_len)
 }
 
+/// An inclusive interval, used for precise integer exhaustiveness checking.
+/// `IntRange`s always store a contiguous range. This means that values are
+/// encoded such that `0` encodes the minimum value for the integer,
+/// regardless of the signedness.
+/// For example, the pattern `-128...127i8` is encoded as `0..=255`.
+/// This makes comparisons and arithmetic on interval endpoints much more
+/// straightforward. See `encode` and `decode` for details.
+struct IntRange<'tcx> {
+    pub range: RangeInclusive<u128>,
+    pub ty: Ty<'tcx>,
+}
+
+impl<'tcx> IntRange<'tcx> {
+    fn from_ctor(tcx: TyCtxt<'_, 'tcx, 'tcx>,
+                 ctor: &Constructor<'tcx>)
+                 -> Option<IntRange<'tcx>> {
+        match ctor {
+            ConstantRange(lo, hi, end) => {
+                assert_eq!(lo.ty, hi.ty);
+                let ty = lo.ty;
+                if let Some(lo) = lo.assert_bits(ty) {
+                    if let Some(hi) = hi.assert_bits(ty) {
+                        // Perform a shift if the underlying types are signed,
+                        // which makes the interval arithmetic simpler.
+                        let (lo, hi) = Self::encode(tcx, ty, (lo, hi));
+                        // Make sure the interval is well-formed.
+                        return if lo > hi || lo == hi && *end == RangeEnd::Excluded {
+                            None
+                        } else {
+                            let offset = (*end == RangeEnd::Excluded) as u128;
+                            Some(IntRange { range: lo..=(hi - offset), ty })
+                        };
+                    }
+                }
+                None
+            }
+            ConstantValue(val) => {
+                let ty = val.ty;
+                if let Some(val) = val.assert_bits(ty) {
+                    let (lo, hi) = Self::encode(tcx, ty, (val, val));
+                    Some(IntRange { range: lo..=hi, ty })
+                } else {
+                    None
+                }
+            }
+            Single | Variant(_) | Slice(_) => {
+                None
+            }
+        }
+    }
+
+    fn encode(tcx: TyCtxt<'_, 'tcx, 'tcx>,
+              ty: Ty<'tcx>,
+              (lo, hi): (u128, u128))
+              -> (u128, u128) {
+        match ty.sty {
+            ty::TyInt(_) => {
+                // FIXME(49937): refactor these bit manipulations into interpret.
+                let bits = tcx.layout_of(ty::ParamEnv::reveal_all().and(ty))
+                              .unwrap().size.bits() as u128;
+                let min = 1u128 << (bits - 1);
+                let mask = !0u128 >> (128 - bits);
+                let offset = |x: u128| x.wrapping_sub(min) & mask;
```

Comment:
Maybe call these `bias` instead of `min` (I just remember). Also note that `add`/`sub` are actually equivalent given the subsequent mask, so maybe.... you just need a XOR with the sign bit, in both of them. 

Maybe have a function called `signed_bias`? Which returns `0` for most types, but `1 << (bits - 1)` for signed ones. And I guess you can use `Integer::from_attr` there too. Then you just `(lo ^ bias, hi ^ bias)` inline wherever you're calling `encode` / `decode` right now.